In [105]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [106]:
!git clone https://github.com/creatorbeen/AlxD_2021_HYOBIN.git

fatal: destination path 'AlxD_2021_HYOBIN' already exists and is not an empty directory.


repository clone from github

In [107]:
import pandas as pd
d = pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube01-Psy.csv")

import 'Youtube01-Psy.csv' file

In [108]:
d.tail()

COMMENT_ID  ... CLASS
345      z13th1q4yzihf1bll23qxzpjeujterydj  ...     0
346  z13fcn1wfpb5e51xe04chdxakpzgchyaxzo0k  ...     0
347    z130zd5b3titudkoe04ccbeohojxuzppvbg  ...     1
348    z12he50arvrkivl5u04cctawgxzkjfsjcc4  ...     1
349  z13vhvu54u3ewpp5h04ccb4zuoardrmjlyk0k  ...     0

[5 rows x 5 columns]

There are more than two columns, but we will only require the content and the calss column. The content column contains the comments and the class column contains the values 1 or 0 for spam or not spam. 

In [109]:
len(d.query('CLASS == 1'))

175

In [110]:
len(d.query('CLASS == 0'))

175

In [111]:
len(d)

350

Count of how many rows in the dataset are spam and how many are not spam. The resulf we acquired is 175 and 175 respectively, which sums up to 350 rows overall in this file.

In [112]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()

In scikit-learn, the bad of words technique is actually called 'CountVectorizer', which means counting how many times each word appears and puts them into a vector. To create a vector, it need to make an object for CountVectorizer, and then perform the fit and transform simultaneously.

In [113]:
dvec = vectorizer.fit_transform(d['CONTENT'])

In [114]:
dvec

<350x1418 sparse matrix of type '<class 'numpy.int64'>'
	with 4354 stored elements in Compressed Sparse Row format>

The performed in two different steps. First comes the fit step, where it discovers which words are present in the dataset, and second is the transform step, which gives you the bag of words matrix for those phrases. The resuld obtained in that matrix is 350 rows by 1418 columns. There are 350 rows, which means it has 350 different comments and 1418 words. 1418 word apparently are word that appear across all of those phrases.

In [115]:
analyze = vectorizer.build_analyzer()

In [116]:
print(d['CONTENT'][349])
analyze(d['CONTENT'][349])

The first billion viewed this because they thought it was really cool, the  other billion and a half came to see how stupid the first billion were...


['the',
 'first',
 'billion',
 'viewed',
 'this',
 'because',
 'they',
 'thought',
 'it',
 'was',
 'really',
 'cool',
 'the',
 'other',
 'billion',
 'and',
 'half',
 'came',
 'to',
 'see',
 'how',
 'stupid',
 'the',
 'first',
 'billion',
 'were']

print a single comment and then run the analyzer on that comment so that see how well the phrases breaks it apart. The comment has been printed first and then analyzing it below, which is just to see how it broke it into words.

In [117]:
vectorizer.get_feature_names()

['00',
 '000',
 '02',
 '034',
 '05',
 '08',
 '10',
 '100',
 '100000415527985',
 '10200253113705769',
 '1030',
 '1073741828',
 '11',
 '1111',
 '112720997191206369631',
 '12',
 '123',
 '124',
 '124923004',
 '126',
 '127',
 '13017194',
 '131338190916',
 '1340488',
 '1340489',
 '1340490',
 '1340491',
 '1340492',
 '1340493',
 '1340494',
 '1340499',
 '1340500',
 '1340502',
 '1340503',
 '1340504',
 '1340517',
 '1340518',
 '1340519',
 '1340520',
 '1340521',
 '1340522',
 '1340523',
 '1340524',
 '134470083389909',
 '1415297812',
 '1495323920744243',
 '1496241863981208',
 '1496273723978022',
 '1498561870415874',
 '161620527267482',
 '171183229277',
 '19',
 '19924',
 '1firo',
 '1m',
 '20',
 '2009',
 '2012',
 '2012bitches',
 '2013',
 '2014',
 '201470069872822',
 '2015',
 '2017',
 '210',
 '23',
 '24',
 '24398',
 '243a',
 '279',
 '29',
 '2b',
 '2billion',
 '2x10',
 '300',
 '3000',
 '313327',
 '315',
 '322',
 '33',
 '33gxrf',
 '39',
 '390875584405933',
 '391725794320912',
 '40beuutvu2zkxk4utgpz8k',
 '

can use the vectorizer feature to find out which word the dataset found after vectorizing. The following is the result found after vectorizing where it starts with numbers and ends with regular words.

In [118]:
dshuf = d.sample(frac=1)

shuffle the dataset with fraction 100% that is adding 'frac=1'

In [119]:
d_train = dshuf[:300]
d_test = dshuf[300:]
d_train_att = vectorizer.fit_transform(d_train['CONTENT']) # fit bag-of-words on training set
d_test_att = vectorizer.transform(d_test['CONTENT']) # reuse on testing set
d_train_label = d_train['CLASS']
d_test_label = d_test['CLASS']

Split the dataset into training and testing sets. Assume that the first 300 will be for training, while the latter 50 will be for testing.
'vectorizer.fit_transform(d_train['CONTENT'])' is an important step. At the stage, it will learn the words and also produce the matrix.



In [120]:
d_train_att

<300x1293 sparse matrix of type '<class 'numpy.int64'>'
	with 3830 stored elements in Compressed Sparse Row format>

In [121]:
d_test_att

<50x1293 sparse matrix of type '<class 'numpy.int64'>'
	with 395 stored elements in Compressed Sparse Row format>

The training set now has 300 rows and 1256 different words or columns, and the testing set has 50 rows, but it has the same 1256 columns.

In [122]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=80)

In [123]:
clf.fit(d_train_att, d_train_label)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=80,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

Building of the random forest classifier.
It will be converting this dataset into 80 different trees and fit the training set so that score its performance on the testing set.

In [124]:
clf.score(d_test_att, d_test_label)

0.92

The output of the score that I received is 96%. 

In [125]:
from sklearn.metrics import confusion_matrix
pred_labels = clf.predict(d_test_att)
confusion_matrix(d_test_label, pred_labels)

array([[23,  2],
       [ 2, 23]])

It seems it got confused between spam and not-spam.

In [126]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, d_train_att, d_train_label, cv=5)
# show average score and +/- two standard deviations away (covering 95% of scores)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.96 (+/- 0.02)


Perform a cross validation with five different splits. The output of the score is 97% accuracy.

In [127]:
# load all datasets and combine them
d = pd.concat([pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube01-Psy.csv"),
               pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube02-KatyPerry.csv"),
               pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube03-LMFAO.csv"),
               pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube04-Eminem.csv"),
               pd.read_csv("/content/AlxD_2021_HYOBIN/Week04_Youtube_dataset/Youtube05-Shakira.csv")])

Print all the data.

In [128]:
len(d)

1956

The entire dataset has five different videos with comments, which means all together we have around 2000 rows.

In [129]:
len(d.query('CLASS == 1'))

1005

In [130]:
len(d.query('CLASS == 0'))

951

On checking all the comments, it noticed that there are 1005 spam comments and 951 not-spam comments, that quite close enough to split it in to even parts.

In [131]:
dshuf = d.sample(frac=1)
d_content = dshuf['CONTENT']
d_label = dshuf['CLASS']

shuffle the entire dataset and separate the comments and the answers.

In [132]:
# set up a pipeline
from sklearn.pipeline import Pipeline, make_pipeline
pipeline = Pipeline([
    ('bag-of-words', CountVectorizer()),
    ('random forest', RandomForestClassifier()),
])
pipeline

Pipeline(memory=None,
         steps=[('bag-of-words',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabu...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

Here a couple of steps with CountVectorizer followed by the random forest. Pipeline is really convenient and will bring together two or more steps so that all the steps are treated as one. The code will build a pipeline with the bad of words, and then use CountVectorizer followed by the random forest classifier. 

In [133]:
# or: pipeline = make_pipeline(CountVectorizer(), RandomForestClassifier())
make_pipeline(CountVectorizer(), RandomForestClassifier())

Pipeline(memory=None,
         steps=[('countvectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, voc...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

Let the pipeline name of each step by itself by adding CountVectorizer in our RandomForestClassifier and it will name them CountVectorizer and RandomForestClassifier.

In [134]:
pipeline.fit(d_content[:1500],d_label[:1500])

Pipeline(memory=None,
         steps=[('bag-of-words',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabu...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

Once the pipeline is created you can just call it fit and it will perform the rest that is first it perform the fit and then transform with the CountVectorizer, followed by a fit with the RandomForestClassifier. That's the benefit of having a pipeline.

In [135]:
pipeline.score(d_content[1500:], d_label[1500:])

0.9539473684210527

This whole procedure will produce a score of about 96. It can only predict a single example with the pipeline.

In [136]:
pipeline.predict(["what a neat video!"])

array([0])

We can only predict a single example with the pipeline. For example, imagine we have a new comment after the dataset has been trained, and we want to know whether the user has just typed this comment or whether it's spam. It's spam cuase '0'

In [137]:
pipeline.predict(["plz subscribe to my channel"])

array([1])

It is not-spam cause '1'

In [138]:
scores = cross_val_score(pipeline, d_content, d_label, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.96 (+/- 0.02)


To overcome this and deploy this classifier into an environment and predict whether it is a spam or not when someone types a new comment. We will use our pipeline to figure out how accurate our cross-validation was. We find in this case that the average accuracy was about 96.

In [139]:
# add tfidf
from sklearn.feature_extraction.text import TfidfTransformer
pipeline2 = make_pipeline(CountVectorizer(),
                          TfidfTransformer(norm=None),
                          RandomForestClassifier())

In [140]:
scores = cross_val_score(pipeline2, d_content, d_label, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.96 (+/- 0.02)


We have produced the counts, we can then produce a TF-IDF score for these counts. We will add this in the pipeline and perform another cross-validation check with the same accuracy(96).

In [141]:
pipeline2.steps

[('countvectorizer',
  CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                  dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                  lowercase=True, max_df=1.0, max_features=None, min_df=1,
                  ngram_range=(1, 1), preprocessor=None, stop_words=None,
                  strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                  tokenizer=None, vocabulary=None)),
 ('tfidftransformer',
  TfidfTransformer(norm=None, smooth_idf=True, sublinear_tf=False, use_idf=True)),
 ('randomforestclassifier',
  RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                         criterion='gini', max_depth=None, max_features='auto',
                         max_leaf_nodes=None, max_samples=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, 

This show the steps required for the pipeline.

In [142]:
# parameter search
parameters = {
    'countvectorizer__max_features': (None, 1000, 2000),
    'countvectorizer__ngram_range': ((1, 1), (1, 2)),  # unigrams or bigrams
    'countvectorizer__stop_words': ('english', None),
    'tfidftransformer__use_idf': (True, False), # effectively turn on/off tfidf
    'randomforestclassifier__n_estimators': (20, 50, 100)
}
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(pipeline2, parameters, n_jobs=-1, verbose=1)

There's another feature of scikit-learn available that allows us to search all of these parameters. 
We can make a little dictionary where we say the name of the pipeline step and then mention what the parameter name would be and this gives us our options. For demonstration, we are going to try maximum number of words or maybe just a maximum of 1,000 or 2,000 words.  Using ngrams, we can mention just single words or pairs of words that are stop words, use the English dictionary of stop words, or don't use stop words, which means in the first case we need to get rid of common words, and in the second case we do not get rid of common words. Using TF-IDF, we use JEG to state whether it's yes or no. The random forest we created uses 20, 50, or 100 trees. Using this, we can perform a grid search, which runs through all of the combinations of parameters and finds out what the best combination is. So, let's give our pipeline number 2, which has the TF-IDF along with it.

In [143]:
grid_search.fit(d_content, d_label)

Fitting 5 folds for each of 72 candidates, totalling 360 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   12.5s
[Parallel(n_jobs=-1)]: Done 196 tasks      | elapsed:  1.0min
[Parallel(n_jobs=-1)]: Done 360 out of 360 | elapsed:  1.8min finished


GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('countvectorizer',
                                        CountVectorizer(analyzer='word',
                                                        binary=False,
                                                        decode_error='strict',
                                                        dtype=<class 'numpy.int64'>,
                                                        encoding='utf-8',
                                                        input='content',
                                                        lowercase=True,
                                                        max_df=1.0,
                                                        max_features=None,
                                                        min_df=1,
                                                        ngram_range=(1, 1),
                                              

fit to perform the search.

In [144]:
print("Best score: %0.3f" % grid_search.best_score_)
print("Best parameters set:")
best_parameters = grid_search.best_estimator_.get_params()
for param_name in sorted(parameters.keys()):
    print("\t%s: %r" % (param_name, best_parameters[param_name]))

Best score: 0.960
Best parameters set:
	countvectorizer__max_features: None
	countvectorizer__ngram_range: (1, 1)
	countvectorizer__stop_words: None
	randomforestclassifier__n_estimators: 100
	tfidftransformer__use_idf: True


Since there is a large number of words, it takes a little while, around 40 seconds, and ultimately finds the best parameters.
It shows nearly 96% accuracy.It used around 2000 words, only single words, used yes to get rid of stop words, had 100 trees in the random forest, and used yes and the IDF and the TF-IDF computation.

#4.2
The outputs are correct except the last output. It might be predict spam that include this word: 'channel'
visit, money, editor, something else are not-spam.

In [145]:
pipeline.predict(["Is it true? Plz take my money angels!"])

array([0])

In [146]:
pipeline.predict(["OMG! This Episode is going to be Awesome"])

array([0])

In [147]:
pipeline.predict(["The editor is the most patience person in the world."])

array([0])

In [148]:
pipeline.predict(["Plz visit my channel and take some money!"])

array([1])

In [149]:
pipeline.predict(["I visit there some weeks ago, but doesn't it."])

array([0])

In [150]:
pipeline.predict(["This channel provide useful information to me."])

array([1])

#4.3
It shows nearly 96% accuracy.It used around 1000 words, only single words, used yes to get rid of stop words, had 100 trees in the random forest, and used yes and the IDF and the TF-IDF computation.

In [151]:
# parameter search
parameters = {
    'countvectorizer__max_features': (None, 1000, 2000),
    'countvectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),  # unigrams, bigrams or trigrams
    'countvectorizer__stop_words': ('english', None),
    'tfidftransformer__use_idf': (True, False), # effectively turn on/off tfidf
    'randomforestclassifier__n_estimators': (20, 50, 100)
}
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(pipeline2, parameters, n_jobs=-1, verbose=1)

In [152]:
grid_search.fit(d_content, d_label)

Fitting 5 folds for each of 108 candidates, totalling 540 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   11.4s
[Parallel(n_jobs=-1)]: Done 196 tasks      | elapsed:  1.4min
[Parallel(n_jobs=-1)]: Done 446 tasks      | elapsed:  2.5min
[Parallel(n_jobs=-1)]: Done 540 out of 540 | elapsed:  3.0min finished


GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('countvectorizer',
                                        CountVectorizer(analyzer='word',
                                                        binary=False,
                                                        decode_error='strict',
                                                        dtype=<class 'numpy.int64'>,
                                                        encoding='utf-8',
                                                        input='content',
                                                        lowercase=True,
                                                        max_df=1.0,
                                                        max_features=None,
                                                        min_df=1,
                                                        ngram_range=(1, 1),
                                              

In [153]:
print("Best score: %0.3f" % grid_search.best_score_)
print("Best parameters set:")
best_parameters = grid_search.best_estimator_.get_params()
for param_name in sorted(parameters.keys()):
    print("\t%s: %r" % (param_name, best_parameters[param_name]))

Best score: 0.960
Best parameters set:
	countvectorizer__max_features: 2000
	countvectorizer__ngram_range: (1, 1)
	countvectorizer__stop_words: 'english'
	randomforestclassifier__n_estimators: 100
	tfidftransformer__use_idf: True


#4.4


In [160]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 1. cv_results_ dictionary -> dataFrame
results = pd.DataFrame(grid_search.cv_results_)


# 2. first 5 rows
results.head(5)



mean_fit_time  std_fit_time  ...  std_test_score  rank_test_score
0       0.237140      0.006107  ...        0.008203               21
1       0.230162      0.005962  ...        0.013198               82
2       0.481381      0.016762  ...        0.007677               10
3       0.463234      0.004455  ...        0.008063               24
4       0.882381      0.007113  ...        0.008390               25

[5 rows x 18 columns]